File: Test/Model3D/Material3D.ts
// @preview-file on clear
import {
	App,
	Camera3D,
	Color,
	Color3,
	Content,
	DirectionalLight3D,
	Director,
	MaterialAlphaMode3D,
	Model3D,
	Vec2,
	Vec3,
	threadLoop,
} from "Dora";
import {SetCond, WindowFlag} from "ImGui";
import * as ImGui from "ImGui";

const view = Director.entry;
const camera = Camera3D();
camera.lookAt(Vec3(0, 2.2, 8), Vec3(0, 1, 0));
Director.pushCamera(camera);
view.setEnvironmentMap("");
view.setEnvironmentIntensity(0.2, 0, 1);

const light = DirectionalLight3D();
light.color = Color3(0xffffff);
light.intensity = 4;
light.angleX = -35;
light.angleY = 25;
view.addChild(light);

const file = "Test/Model3D/Assets/Model/Duck.glb";
const changed = Model3D(file);
changed.position = Vec3(-1.5, 0, 0);
view.addChild(changed);

const original = Model3D(file);
original.position = Vec3(1.5, 0, 0);
view.addChild(original);

const changedMaterial = changed.getMaterial(0);
const originalMaterial = original.getMaterial(0);
if (!changedMaterial || !originalMaterial) {
	error("Duck material slot 0 is missing");
}

const originalColor = originalMaterial.baseColor.toARGB();
const originalMetallic = originalMaterial.metallic;
const originalRoughness = originalMaterial.roughness;

changedMaterial.baseColor = Color(0xffff4040);
changedMaterial.emissive = Color3(0x220000);
changedMaterial.metallic = 1;
changedMaterial.roughness = 0.15;
changedMaterial.alphaMode = MaterialAlphaMode3D.Opaque;
changedMaterial.alphaCutoff = 0.45;
changedMaterial.clearNormalTexture();

const copyOnWriteValid = changed.materialCount === 1
	&& original.materialCount === 1
	&& changedMaterial.baseColor.toARGB() === 0xffff4040
	&& Math.abs(changedMaterial.metallic - 1) < 0.001
	&& Math.abs(changedMaterial.roughness - 0.15) < 0.001
	&& originalMaterial.baseColor.toARGB() === originalColor
	&& Math.abs(originalMaterial.metallic - originalMetallic) < 0.001
	&& Math.abs(originalMaterial.roughness - originalRoughness) < 0.001;

let elapsed = 0;
let status = "Pending";
threadLoop(() => {
	elapsed += App.deltaTime;
	if (status === "Pending" && elapsed >= 1.5) {
		status = copyOnWriteValid ? "PASS" : "FAIL";
		const screenshot = App.saveScreenshot("/tmp/dora-3d-material/material-runtime");
		const summary = `MATERIAL3D_SUMMARY status=${status} slots=${changed.materialCount} changedColor=${changedMaterial.baseColor.toARGB()} originalColor=${originalMaterial.baseColor.toARGB()} changedMetallic=${changedMaterial.metallic.toFixed(2)} originalMetallic=${originalMaterial.metallic.toFixed(2)} changedRoughness=${changedMaterial.roughness.toFixed(2)} originalRoughness=${originalMaterial.roughness.toFixed(2)} screenshot=${screenshot}`;
		Content.save("/tmp/dora-3d-material/summary.txt", summary);
		print(summary);
	}

	ImGui.SetNextWindowPos(Vec2(12, 12), SetCond.Always);
	ImGui.SetNextWindowSize(Vec2(470, 0), SetCond.Always);
	ImGui.SetNextWindowBgAlpha(0.8);
	ImGui.Begin("Material3D Copy-on-write", [WindowFlag.NoSavedSettings, WindowFlag.NoFocusOnAppearing], () => {
		ImGui.Text(`Material slots: ${changed.materialCount}`);
		ImGui.Text(`Changed: metallic ${changedMaterial.metallic.toFixed(2)}, roughness ${changedMaterial.roughness.toFixed(2)}`);
		ImGui.Text(`Original: metallic ${originalMaterial.metallic.toFixed(2)}, roughness ${originalMaterial.roughness.toFixed(2)}`);
		ImGui.Text("Left instance changed, right instance unchanged");
		ImGui.Text(`Verification: ${status}`);
	});
	return false;
});

File: Test/Model3D/Material3D.lua
-- [ts]: Material3D.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__NumberToFixed = ____lualib.__TS__NumberToFixed -- 1
local ____exports = {} -- 1
local ____Dora = require("Dora") -- 2
local App = ____Dora.App -- 3
local Camera3D = ____Dora.Camera3D -- 4
local Color = ____Dora.Color -- 5
local Color3 = ____Dora.Color3 -- 6
local Content = ____Dora.Content -- 7
local DirectionalLight3D = ____Dora.DirectionalLight3D -- 8
local Director = ____Dora.Director -- 9
local Model3D = ____Dora.Model3D -- 11
local Vec2 = ____Dora.Vec2 -- 12
local Vec3 = ____Dora.Vec3 -- 13
local threadLoop = ____Dora.threadLoop -- 14
local ImGui = require("ImGui") -- 17
local view = Director.entry -- 19
local camera = Camera3D() -- 20
camera:lookAt( -- 21
	Vec3(0, 2.2, 8), -- 21
	Vec3(0, 1, 0) -- 21
) -- 21
Director:pushCamera(camera) -- 22
view:setEnvironmentMap("") -- 23
view:setEnvironmentIntensity(0.2, 0, 1) -- 24
local light = DirectionalLight3D() -- 26
light.color = Color3(16777215) -- 27
light.intensity = 4 -- 28
light.angleX = -35 -- 29
light.angleY = 25 -- 30
view:addChild(light) -- 31
local file = "Test/Model3D/Assets/Model/Duck.glb" -- 33
local changed = Model3D(file) -- 34
changed.position = Vec3(-1.5, 0, 0) -- 35
view:addChild(changed) -- 36
local original = Model3D(file) -- 38
original.position = Vec3(1.5, 0, 0) -- 39
view:addChild(original) -- 40
local changedMaterial = changed:getMaterial(0) -- 42
local originalMaterial = original:getMaterial(0) -- 43
if not changedMaterial or not originalMaterial then -- 43
	error("Duck material slot 0 is missing") -- 45
end -- 45
local originalColor = originalMaterial.baseColor:toARGB() -- 48
local originalMetallic = originalMaterial.metallic -- 49
local originalRoughness = originalMaterial.roughness -- 50
changedMaterial.baseColor = Color(4294918208) -- 52
changedMaterial.emissive = Color3(2228224) -- 53
changedMaterial.metallic = 1 -- 54
changedMaterial.roughness = 0.15 -- 55
changedMaterial.alphaMode = 0 -- 56
changedMaterial.alphaCutoff = 0.45 -- 57
changedMaterial:clearNormalTexture() -- 58
local copyOnWriteValid = changed.materialCount == 1 and original.materialCount == 1 and changedMaterial.baseColor:toARGB() == 4294918208 and math.abs(changedMaterial.metallic - 1) < 0.001 and math.abs(changedMaterial.roughness - 0.15) < 0.001 and originalMaterial.baseColor:toARGB() == originalColor and math.abs(originalMaterial.metallic - originalMetallic) < 0.001 and math.abs(originalMaterial.roughness - originalRoughness) < 0.001 -- 60
local elapsed = 0 -- 69
local status = "Pending" -- 70
threadLoop(function() -- 71
	elapsed = elapsed + App.deltaTime -- 72
	if status == "Pending" and elapsed >= 1.5 then -- 72
		status = copyOnWriteValid and "PASS" or "FAIL" -- 74
		local screenshot = App:saveScreenshot("/tmp/dora-3d-material/material-runtime") -- 75
		local summary = (((((((((((((((("MATERIAL3D_SUMMARY status=" .. status) .. " slots=") .. tostring(changed.materialCount)) .. " changedColor=") .. tostring(changedMaterial.baseColor:toARGB())) .. " originalColor=") .. tostring(originalMaterial.baseColor:toARGB())) .. " changedMetallic=") .. __TS__NumberToFixed(changedMaterial.metallic, 2)) .. " originalMetallic=") .. __TS__NumberToFixed(originalMaterial.metallic, 2)) .. " changedRoughness=") .. __TS__NumberToFixed(changedMaterial.roughness, 2)) .. " originalRoughness=") .. __TS__NumberToFixed(originalMaterial.roughness, 2)) .. " screenshot=") .. screenshot -- 76
		Content:save("/tmp/dora-3d-material/summary.txt", summary) -- 77
		print(summary) -- 78
	end -- 78
	ImGui.SetNextWindowPos( -- 81
		Vec2(12, 12), -- 81
		"Always" -- 81
	) -- 81
	ImGui.SetNextWindowSize( -- 82
		Vec2(470, 0), -- 82
		"Always" -- 82
	) -- 82
	ImGui.SetNextWindowBgAlpha(0.8) -- 83
	ImGui.Begin( -- 84
		"Material3D Copy-on-write", -- 84
		{"NoSavedSettings", "NoFocusOnAppearing"}, -- 84
		function() -- 84
			ImGui.Text("Material slots: " .. tostring(changed.materialCount)) -- 85
			ImGui.Text((("Changed: metallic " .. __TS__NumberToFixed(changedMaterial.metallic, 2)) .. ", roughness ") .. __TS__NumberToFixed(changedMaterial.roughness, 2)) -- 86
			ImGui.Text((("Original: metallic " .. __TS__NumberToFixed(originalMaterial.metallic, 2)) .. ", roughness ") .. __TS__NumberToFixed(originalMaterial.roughness, 2)) -- 87
			ImGui.Text("Left instance changed, right instance unchanged") -- 88
			ImGui.Text("Verification: " .. status) -- 89
		end -- 84
	) -- 84
	return false -- 91
end) -- 71
return ____exports -- 71